### PR TITLE
Document NonZeroXxx layout guarantees

### PR DESCRIPTION
**Pull request by joshlf**
Document that `NonZeroXxx` has the same layout and bit validity as `Xxx` with the exception of `0`.

### CODE REVIEW

**Comment by joshtriplett** [comment]
I feel like this was always the commitment, but at the same time, better to confirm that consensus rather than assuming it.

@rust-lang/libs-api, are we comfortable committing to the guarantee that NonZeroX is compatible with X?

**Comment by joshtriplett** [comment]
@rfcbot merge

**Comment by joshtriplett** [comment]
@rfcbot merge

**Comment by dtolnay** [comment]
Do we further want to guarantee that it has the same ABI as X, for example that `extern "C" fn() -> NonZeroU8` is bindable to/from C `uint8_t (*)(void)`? That isn't covered by what's written in the PR.

**Comment by joshlf** [comment]
> Seems okay but do we further want to guarantee that it has the same ABI as X, for example that `extern "C" fn() -> NonZeroU8` is bindable to/from C `uint8_t(void)`? That isn't covered by what's written in the PR.

Presumably we'd need to clarify that 0 is an exception? It's not safe in the general case to use `uint8_t(void)` as `extern "C" fn() -> NonZeroU8` (since the `uint8_t(void)` might return 0), nor is it safe in the general case to use `extern "C" fn(NonZeroU8)` as `void(uint8_t)` (since the `void(uint8_t)` might be passed 0).

**Comment by dtolnay** [comment]
Committing to the ABI being the same isn't about saying it's safe/correct to arbitrarily interchange the two signatures. It's just about whether the registers involved in the calling convention are guaranteed to line up, e.g. that if `fn() -> u8` returns its result in a particular register then `fn() -> NonZeroU8` also returns in that same register. This is something you need to guarantee and is not the case in some "obvious" cases, for example that `std::unique_ptr<T> ()` and `T* ()` do *not* return the same way, even though unique_ptr is "just a pointer" and has the same layout as a pointer.

**Comment by wwylele** [comment]
>... is guaranteed to have the same layout and **bit validity** as ...

Isn't the bit validity part false, because all 0 bits are not valid for NonZeroXxx?

**Comment by joshtriplett** [comment]
> Seems okay but do we further want to guarantee that it has the same ABI as X, for example that `extern "C" fn() -> NonZeroU8` is bindable to/from C `uint8_t(void)`? That isn't covered by what's written in the PR.

I think it was intended to be, but making that more explicit would help. We should document that it is repr(transparent) and can be used in FFI.

**Comment by m-ou-se** [comment]
@rfcbot concern implications

I think it's important to also say that this doesn't mean that e.g. `Option<NonZeroU8>` and `Option<u8>` have the same layout. Saying that `NonZeroU8` and `u8` have the same layout might seem to imply that they always take the same amount of space, even when combined with other things.

**Comment by workingjubilee** [comment]
I agree that clarifying the differences between `Option<u8>`, `Option<NZu8>`, and `u8` is important. In particular, while I believe that they have the same size, and I believe `NZu8` has the same ABI as u8, `Option<NZu8>` and `u8` would not necessarily be ABI interchangeable even if `NZu8` and `u8` are. It is very, very plausible to me that someone gets this wrong as a result.

**Comment by dtolnay** [comment]
As far as I know:

- `Option<NonZeroU8>` guarantees same **layout** as `u8` since #51396.
- `NonZeroU8` guarantees same **layout** as `u8` once the FCP here has gone through.
- IMO `NonZeroU8` should further guarantee same **ABI** as `u8`, but I wouldn't block on this and it can happen later.
- `Option<NonZeroU8>` continues to not have a specific guaranteed **ABI**. Whether it's the same as C `uint8_t` or as C `struct S { uint8_t x; };` (which are not the same as one another) has never been settled.

File: library/core/src/num/nonzero.rs
Comment (joshtriplett) [inline library/core/src/num/nonzero.rs:41]:
```suggestion
            /// with the exception that `0` is not a valid instance.
            #[doc = concat!("`Option<", stringify!($Ty), "` is guaranteed to be compatible with `", stringify!($Int), "`,"
            /// including in FFI.
```

**Comment by joshtriplett** [comment]
@m-ou-se Does https://github.com/rust-lang/rust/pull/94786#discussion_r882043136 address your concern?

**Comment by JohnCSimon** [comment]
triage:
@m-ou-se :point_up: ?

**Comment by m-ou-se** [comment]
@rfcbot resolve implications

**Comment by joshlf** [comment]
Is there some manual intervention needed to get this merged?

**Comment by JohnCSimon** [comment]
Ping from triage: 
@joshlf 
FYI: when a PR is ready for review, post a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

**Comment by joshlf** [comment]
Thanks, @JohnCSimon!

@rustbot ready

**Review by dtolnay — APPROVED** [review]

**Comment by joshlf** [comment]
@joshtriplett I've realized that I may have prematurely accepted your suggested language that made it into the merge:

```rust
#[doc = concat!("`Option<", stringify!($Ty), ">` is guaranteed to be compatible with `", stringify!($Int), "`,")]
/// including in FFI.
```

Does "compatible with" have a technical definition? In https://github.com/rust-lang/rust/pull/104082, there's discussion of using the language "layout and bit validity", although I'm not sure whether that implicitly provides the same guarantees as your "...including in FFI" language.

Sorry for only bringing this up after the merge! Whatever the outcome of this discussion is, I can include any proposed changes in https://github.com/rust-lang/rust/pull/104082.

**Pull request merged**
